### PR TITLE
Update /instrument flag values

### PR DESCRIPTION
**Pull request by drognanar**
Make instrument

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 1]
```diff
@@ -109,7 +109,7 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
             string runtimeMetadataVersion = null;
             bool errorEndLocation = false;
             bool reportAnalyzer = false;
-            string instrument = "";
+            InstrumentationKind instrument = InstrumentationKind.None;
```
Comment (tmat) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: var

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 2]
```diff
@@ -32,10 +32,9 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
         public bool IncludePrivateMembers { get; private set; }
 
         /// <summary>
-        /// If not empty then compiler should enable instrumentation.
-        /// The string specifies instrumentation settings.
+        /// Type of instrumentation that should be added to the output binary.
         /// </summary>
-        public string Instrument { get; private set; }
+        public InstrumentationKind Instrument { get; private set; }
```
Comment (tmat) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: I think the property should be named "InstrumentationKind". It's odd to use a verb.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 3]
```diff
@@ -377,7 +379,7 @@ public EmitOptions WithIncludePrivateMembers(bool value)
             return new EmitOptions(this) { IncludePrivateMembers = value };
         }
 
-        public EmitOptions WithInstrument(string instrument)
+        public EmitOptions WithInstrument(InstrumentationKind instrument)
```
Comment (tmat) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: ```WithInstrumentation``` or ```WithInstrumentationKind```

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 4]
```diff
@@ -4428,8 +4428,8 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
  /optimize[+|-]                Enable optimizations (Short form: /o)
  /deterministic                Produce a deterministic assembly
                                (including module version GUID and timestamp)
- /instrument:options           Sets instrumentation options.
-                               If empty instrumentation is disabled.
+ /instrument:{TestCoverage}    Specify instrumentation method.
+                               If flag is not passed then instrumentation is disabled.
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: I'm not sure precisely what this means. Is "flag" "instrument" or "TestCoverage"?

I don't know if there is a way to word this such that adding a new kind of instrumentation involves only adding text.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 5]
```diff
@@ -4936,4 +4936,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_SourceLinkRequiresPortablePdb" xml:space="preserve">
     <value>/sourcelink switch is only supported when emitting Portable PDB (/debug:portable or /debug:embedded must be specified).</value>
   </data>
+  <data name="ERR_BadInstrumentationKind" xml:space="preserve">
+    <value>Invalid option '{0}' for /instrument; must be testcoverage only</value>
+  </data>
+  <data name="ERR_InvalidInstrumentationKind" xml:space="preserve">
+    <value>Invalid instrumentation kind: {0}</value>
+  </data>
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Are both of the above necessary?

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 6]
```diff
@@ -109,7 +109,7 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
             string runtimeMetadataVersion = null;
             bool errorEndLocation = false;
             bool reportAnalyzer = false;
-            string instrument = "";
+            HashSet<string> instrument = new HashSet<string>();
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Should the hash set be configured to do case-insensitive comparison?

**Comment by JohnHamby** [comment]
At a quick glance this looks fine to me, modulo the couple of questions I asked (which I didn't repeat for the corresponding VB code).

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 7]
```diff
@@ -32,10 +33,9 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
         public bool IncludePrivateMembers { get; private set; }
 
         /// <summary>
-        /// If not empty then compiler should enable instrumentation.
-        /// The string specifies instrumentation settings.
+        /// Type of instrumentation that should be added to the output binary.
         /// </summary>
-        public string Instrument { get; private set; }
+        public HashSet<string> Instrument { get; private set; }
```
Comment (tmat) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: >HashSet<string> [](start = 15, length = 15)

Why not enum flags?

Comment (tmat) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: This is certainly wrong since HashSet is mutable. I'd prefer not use strings. If we use strings then ImmutableArray<string> would be more appropriate.

---
In reply to: [73979817](https://github.com/dotnet/roslyn/pull/13013#discussion_r73979817) [](ancestors = 73979817)

Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: A flags enum is limited to 32 (or 64) values. An immutable array of non-flags enum values probably works.

Comment (tmat) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Are we gonna have more than 64 kinds of instrumentation? If that ever happens (I doubt it will) we can add another enum, or even immutable array later.

**Comment by drognanar** [comment]
FYI @dotnet/roslyn-compiler

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 8]
```diff
@@ -1426,5 +1426,13 @@ internal enum ErrorCode
         ERR_TypeInferenceFailedForImplicitlyTypedOutVariable = 8197,
         ERR_ExpressionTreeContainsOutVariable = 8198,
         #endregion diagnostics for out var
+
+        // gap here; available 8149-8499
+
+        #region diagnostics for instrumentation
+
+        ERR_InvalidInstrumentationKind = 8501,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: Do you expect to have bunch of new error codes for instrumentation?
If not, we just compacted the C#7 error codes, so we'd rather not introduce new large gaps. You could grab `8111`.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 9]
```diff
@@ -177,6 +178,14 @@ public bool Equals(EmitOptions other)
                 return false;
             }
 
+            for (int i = 0; i < this.InstrumentationKinds.Length; i++)
+            {
+                if (this.InstrumentationKinds[i] != other.InstrumentationKinds[i])
+                {
+                    return false;
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: It looks like order matters in this equality comparison. Can you confirm this is intentional?

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 10]
```diff
@@ -111,7 +112,7 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
             string runtimeMetadataVersion = null;
             bool errorEndLocation = false;
             bool reportAnalyzer = false;
-            string instrument = "";
+            ArrayBuilder<InstrumentationKind> instrument = ArrayBuilder<InstrumentationKind>.GetInstance();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: In the past John said there would be an unbounded number of such kinds, but I still don't understand the concept of different instrumentation kinds.
Would you have an example of two different "kinds" (to clarify the concept) which can be simultaneously be instrumented (to justify why we need an array of them).

File: src/Compilers/Core/Portable/Emit/InstrumentationKind.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.Emit
+{
+    public enum InstrumentationKind
+    {
+        None = 0,
+        TestCoverage = 1,
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Emit/InstrumentationKind.cs]: Previously, @JohnHamby had indicated that instrumentation kinds would not be an enum. I'm confused.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 12]
```diff
@@ -225,6 +239,14 @@ internal void ValidateOptions(DiagnosticBag diagnostics, CommonMessageProvider m
                 diagnostics.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_InvalidDebugInformationFormat, Location.None, (int)DebugInformationFormat));
             }
 
+            foreach (var instrumentationKind in InstrumentationKinds)
+            {
+                if (!instrumentationKind.IsValid())
+                {
+                    diagnostics.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_InvalidInstrumentationKind, Location.None, (int)instrumentationKind));
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Clarifying the concept of instrumentation kind will also help me understand if/whether a numeric index is a meaningful way to represent it and communicate it to users. For instance, should an instrumentation kind have a number or an identifier/name?

Will this error message be useful? "Invalid instrumentation kind: 42"

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 13]
```diff
@@ -245,7 +267,7 @@ internal void ValidateOptions(DiagnosticBag diagnostics, CommonMessageProvider m
             }
         }
 
-        internal bool EmitDynamicAnalysisData => !string.IsNullOrEmpty(Instrument);
+        internal bool EmitDynamicAnalysisData => !InstrumentationKinds.IsDefaultOrEmpty && InstrumentationKinds.Contains(InstrumentationKind.TestCoverage);
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: If I understood correctly, it seems that `InstrumentationKinds` cannot be default (it could either be empty or be filled).

File: src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb [hunk 14]
```diff
@@ -272,6 +272,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
+        Public Overrides ReadOnly Property ERR_InvalidInstrumentationKind As Integer
+            Get
+                Return ERRID.ERR_InvalidInstrumentationKind
+            End Get
+        End Property
+
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb]: Nit: extra empty line here

File: src/Compilers/Core/Portable/Emit/InstrumentationKind.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Emit/InstrumentationKind.cs]: The `using` is probably not required.

**Comment by jcouv** [comment]
Done with review pass. No major concern, mostly some clarifying design questions.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 12 above)
Comment (tmat) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: I think this is ok, since the user needs to cast int to the enum type to produce invalid value. All valid values are named (have a enum member). Invalid values are unnamed (don't have enum value), so there is nothing better to display.

File: src/Compilers/Core/Portable/Emit/InstrumentationKind.cs (hunk 11 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/InstrumentationKind.cs]: I don't have a strong opinion about whether instrumentation kinds are encoded using an enum or as strings--whatever gets through review is fine with me.

An advantage of an enum is that the allowed values are explicit, and that's my preference, but mostly we need something that works merged as soon as practical.

**Comment by jcouv** [comment]
LGTM

**Comment by ManishJayaswal** [comment]
@AlekseyTs for CR signoff.

File: src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs [hunk 16]
```diff
@@ -73,7 +73,7 @@ public string ChecksumAlgorithm
         }
         
         /// <summary>
-        /// An instrument flag that specifies instrumentation settings.
+        /// An instrument flag that specifies instrumentation kind.
```
Comment (gafter) [inline src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs]: It is hard to tell whether this is intended to support a single kind of instrumentation, or multiple kinds. I does not help that this sentence is missing the article "the" (e.g. "the instrumentation kind") and also "kind" is singular, making it somewhat awkward and ambiguous.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 17]
```diff
@@ -377,14 +399,19 @@ public EmitOptions WithIncludePrivateMembers(bool value)
             return new EmitOptions(this) { IncludePrivateMembers = value };
         }
 
-        public EmitOptions WithInstrument(string instrument)
+        public EmitOptions WithInstrument(params InstrumentationKind[] instrument)
+        {
+            return WithInstrument(ImmutableArray.Create(instrument));
+        }
+
+        public EmitOptions WithInstrument(ImmutableArray<InstrumentationKind> instrument)
```
Comment (gafter) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Should be WithInstrumentationKinds.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 18]
```diff
@@ -377,14 +399,19 @@ public EmitOptions WithIncludePrivateMembers(bool value)
             return new EmitOptions(this) { IncludePrivateMembers = value };
         }
 
-        public EmitOptions WithInstrument(string instrument)
+        public EmitOptions WithInstrument(params InstrumentationKind[] instrument)
```
Comment (gafter) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: >WithInstrument [](start = 27, length = 14)

Please remove this from the API. We generally have only the single overload with an immutable input for the With methods.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 19]
```diff
@@ -133,7 +134,7 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
             string runtimeMetadataVersion = null,
             bool tolerateErrors = false,
             bool includePrivateMembers = false,
-            string instrument = "")
+            ImmutableArray<InstrumentationKind> instrument = default(ImmutableArray<InstrumentationKind>))
```
Comment (gafter) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: >instrument [](start = 48, length = 10)

Should be called `instrumentationKinds`.

**Comment by gafter** [comment]
I'm done reviewing. I just have a couple of public API changes, please.

File: src/Compilers/Core/Portable/Emit/InstrumentationKind.cs [hunk 20]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Emit
+{
+    public enum InstrumentationKind
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/InstrumentationKind.cs]: >InstrumentationKind [](start = 16, length = 19)

Please add doc comments to the type and its members.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 21]
```diff
@@ -32,10 +34,9 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
         public bool IncludePrivateMembers { get; private set; }
 
         /// <summary>
-        /// If not empty then compiler should enable instrumentation.
-        /// The string specifies instrumentation settings.
+        /// Type of instrumentation that should be added to the output binary.
         /// </summary>
-        public string Instrument { get; private set; }
+        public ImmutableArray<InstrumentationKind> InstrumentationKinds { get; private set; }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: >ImmutableArray<InstrumentationKind> [](start = 15, length = 35)

Could this just be a bitmask or do we expect to have more than 32/64 different kinds?

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 22]
```diff
@@ -189,11 +198,17 @@ public bool Equals(EmitOptions other)
                 this.RuntimeMetadataVersion == other.RuntimeMetadataVersion &&
                 this.TolerateErrors == other.TolerateErrors &&
                 this.IncludePrivateMembers == other.IncludePrivateMembers &&
-                this.Instrument == other.Instrument;
+                this.InstrumentationKinds == other.InstrumentationKinds;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: >this.InstrumentationKinds == other.InstrumentationKinds [](start = 16, length = 55)

What does this comparison compare and do we want to do this comparison?

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 23]
```diff
@@ -177,6 +178,14 @@ public bool Equals(EmitOptions other)
                 return false;
             }
 
+            for (int i = 0; i < this.InstrumentationKinds.Length; i++)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: >i < this.InstrumentationKinds.Length [](start = 28, length = 36)

How are we handling different length between this.InstrumentationKinds and other.InstrumentationKinds?

**Comment by AlekseyTs** [comment]
>         public bool Equals(EmitOptions other)

Do we have tests for Equals/GetHashCode with respect to InstrumentationKinds values?

---
Refers to: src/Compilers/Core/Portable/Emit/EmitOptions.cs:174 in 27c2a2f. [](commit_id = 27c2a2fa7c3e55f3c5f4631e713bf0ff58af444d, deletion_comment = False)

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 21 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: We are leaving open the possibility of having more than 64 different kinds.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 24]
```diff
@@ -189,11 +198,17 @@ public bool Equals(EmitOptions other)
...
+                this.InstrumentationKinds == other.InstrumentationKinds;
         }
 
         public override int GetHashCode()
         {
+            var instrumentationKindHash = 0;
+            foreach (var instrumentationKind in InstrumentationKinds)
+            {
+                instrumentationKindHash = Hash.Combine((int)instrumentationKind, instrumentationKindHash);
+            }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Note, we have Hash.CombineValues and SequenceEquals helpers, see how they are used in ParseOptions.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 25]
```diff
@@ -245,7 +267,7 @@ internal void ValidateOptions(DiagnosticBag diagnostics, CommonMessageProvider m
             }
         }
 
-        internal bool EmitDynamicAnalysisData => !string.IsNullOrEmpty(Instrument);
+        internal bool EmitDynamicAnalysisData => InstrumentationKinds.Contains(InstrumentationKind.TestCoverage);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: >EmitDynamicAnalysisData [](start = 22, length = 23)

Perhaps the property should be renamed to EmitTestCoverageData?

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 17 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: +1

---
In reply to: [75935486](https://github.com/dotnet/roslyn/pull/13013#discussion_r75935486) [](ancestors = 75935486)

**Comment by AlekseyTs** [comment]
>         public void TestFieldsForEqualsAndGetHashCode()

I think we need more targeted tests for equality around InstrumentationKinds.

---
Refers to: src/Compilers/Core/CodeAnalysisTest/Emit/EmitOptionsTests.cs:65 in 27c2a2f. [](commit_id = 27c2a2fa7c3e55f3c5f4631e713bf0ff58af444d, deletion_comment = False)

File: src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs [hunk 26]
```diff
@@ -73,7 +73,7 @@ public string ChecksumAlgorithm
...
         /// </summary>
         public string Instrument
```
Comment (AlekseyTs) [inline src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs]: >Instrument [](start = 22, length = 10)

Should this property be renamed? Also, what format the value is using?

File: src/Compilers/Core/MSBuildTaskTests/CscTests.cs [hunk 27]
```diff
@@ -113,8 +113,12 @@ public void ChecksumAlgorithmOption()
         public void InstrumentTestNamesFlag()
```
Comment (AlekseyTs) [inline src/Compilers/Core/MSBuildTaskTests/CscTests.cs]: >InstrumentTestNamesFlag [](start = 20, length = 23)

How does one specify more than one kind?

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 28]
```diff
@@ -2996,39 +2996,53 @@ public void ParseOut2()
         public void ParseInstrumentTestNames()
         {
             var parsedArgs = DefaultParse(SpecializedCollections.EmptyEnumerable<string>(), _baseDirectory);
-            Assert.Equal("", parsedArgs.EmitOptions.Instrument);
+            Assert.Equal(ImmutableHashSet<InstrumentationKind>.Empty, parsedArgs.EmitOptions.InstrumentationKinds);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: >ImmutableHashSet [](start = 25, length = 16)

Is the property an ImmutableArray or an ImmutableHashSet?

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 29]
```diff
@@ -2996,39 +2996,53 @@ public void ParseOut2()
         public void ParseInstrumentTestNames()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: >ParseInstrumentTestNames [](start = 20, length = 24)

It looks like we are not testing scenarios when /instrument switch is specified multiple times.

File: src/Dependencies/PooledObjects/Microsoft.CodeAnalysis.PooledObjects.shproj [hunk 30]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup Label="Globals">
-    <ProjectGuid>c1930979-c824-496b-a630-70f5369a636f</ProjectGuid>
+    <ProjectGuid>cd77a8cc-2885-47a7-b99c-fbf13353400b</ProjectGuid>
```
Comment (AlekseyTs) [inline src/Dependencies/PooledObjects/Microsoft.CodeAnalysis.PooledObjects.shproj]: ><ProjectGuid>cd77a8cc-2885-47a7-b99c-fbf13353400b</ProjectGuid> [](start = 4, length = 63)

Is this change necessary?

**Comment by AlekseyTs** [comment]
I am done with review pass.

**Comment by drognanar** [comment]
Updated PR.
Added a test validating GetHash,Equals for InstrumetnationKinds property.
Added a test that parses multiple instrumentation kinds (one of the strings is invalid since there only exists one InstrumentationKind).

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 31]
```diff
@@ -1656,6 +1657,24 @@ private static void ValidateWin32Settings(string win32ResourceFile, string win32
             }
         }
 
+        internal static IEnumerable<InstrumentationKind> ParseInstrumentationKinds(string value, IList<Diagnostic> diagnostics)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: `private`

Comment (cston) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Consider passing in the `ArrayBuilder<InstrumentationKind>` rather than returning `IEnumerable<InstrumentationKind>`. That would avoid an iterator class and would be consistent with `diagnostics` which is an argument.

Comment (cston) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: No need to change from iterator. It looks like that is the pattern for `Parse` methods.

Comment (cston) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Not part of this change but it looks like `ParseAnalyzers` and `ParseAssemblyReferences` can be `static`. Same for `ParseAnalyzers` in VB.

File: src/Compilers/Core/MSBuildTask/Shared/Microsoft.CSharp.Core.targets [hunk 32]
```diff
@@ -95,7 +95,7 @@
           FileAlignment="$(FileAlignment)"
           GenerateFullPaths="$(GenerateFullPaths)"
           HighEntropyVA="$(HighEntropyVA)"
-          Instrument="$(Instrument)"
+          InstrumentationKinds="$(Instrument)"
```
Comment (AlekseyTs) [inline src/Compilers/Core/MSBuildTask/Shared/Microsoft.CSharp.Core.targets]: >Instrument [](start = 34, length = 10)

Should this be renamed too? Also, it looks like things that might have multiple values are using Items rather than a property.

File: src/Compilers/Core/Portable/Emit/InstrumentationKind.cs [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+    /// </summary>
+    public enum InstrumentationKind
+    {
+        /// <summary>
+        /// No instrumentation.
+        /// </summary>
+        None = 0,
+
+        /// <summary>
+        /// Instruments the binary to add....
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/InstrumentationKind.cs]: >Instruments the binary to add.... [](start = 12, length = 33)

It looks like the comment is incomplete.

**Comment by AlekseyTs** [comment]
>         Public Sub ParseInstrumentTestNames()

It looks like we are missing a test with multiple valid values, even duplicate ones. Same true for C#.

---
Refers to: src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb:421 in 74a7879. [](commit_id = 74a78795f598c7e248ff732d9b3b34f9a44b5bd4, deletion_comment = False)

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs [hunk 34]
```diff
@@ -189,11 +190,17 @@ public bool Equals(EmitOptions other)
...
+                this.InstrumentationKinds.SequenceEqual(other.InstrumentationKinds, (a, b) => a == b);
         }
 
         public override int GetHashCode()
         {
+            var instrumentationKindHash = 0;
+            foreach (var instrumentationKind in InstrumentationKinds)
+            {
+                instrumentationKindHash = Hash.Combine((int)instrumentationKind, instrumentationKindHash);
+            }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: Do we still need this code?

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs (hunk 8 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: @drognanar did this. (I'll be adding notes to myself so that I can figure out what remains to be done.)

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 9 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: @drognanar addressed this by using SequenceEqual, which does not require sequences to have equal values in the same order in order test as equal.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 24 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: The foreach loop will be removed.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 9 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: The code looks different now, because @drognanar changed it to use SequenceEqual, but equality of EmitOptions still implies identical order of instrumentation kinds. That isn't too significant now, in that only one instrumentation kind can be specified, but seems questionable. Tagging @gafter @AlekseyTs @cston for opinions.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 13 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: @drognanar addressed this issue.

File: src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb (hunk 14 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb]: @drognanar addressed this issue.

File: src/Compilers/Core/Portable/Emit/InstrumentationKind.cs (hunk 15 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/InstrumentationKind.cs]: @drognanar removed the using.

File: src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs (hunk 16 above)
Comment (JohnHamby) [inline src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs]: @drognanar addressed this, clarifying that multiple instrumentation kinds can appear.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 17 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: @drognanar made the change to WithInstrumentationKinds.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 18 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: @drognanar removed the params variant.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 19 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: @drognanar made the requested name change.

File: src/Compilers/Core/Portable/Emit/InstrumentationKind.cs (hunk 20 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/InstrumentationKind.cs]: @drognanar added doc comments.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 22 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: @drognanar changed the comparison to use SequenceEqual, which is more meaningful here. Therer is still a question about whether equality should allow differences in order of instrumentation kinds.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 23 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: @drognanar addressed this by using SequenceEqual.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 25 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: @drognanar made the requested name change.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs (hunk 28 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: @drognanar restructured the tests in a way that resolves this issue.

File: src/Compilers/Core/Portable/Emit/EmitOptions.cs (hunk 34 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/EmitOptions.cs]: No, this code will be removed.

File: src/Compilers/Core/Portable/Emit/InstrumentationKind.cs (hunk 33 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Emit/InstrumentationKind.cs]: I think the intent was that the name appearing in the code complete the comment. Since this has confused at least one reviewer, I'll make the comment self contained, with apologies to @drognanar .

File: src/Compilers/Core/MSBuildTask/Shared/Microsoft.CSharp.Core.targets (hunk 32 above)
Comment (JohnHamby) [inline src/Compilers/Core/MSBuildTask/Shared/Microsoft.CSharp.Core.targets]: After lots of back and forth, the current deliberate decision is that Instrument is a property. There is precedent for multiple items being combined in a property, e.g. NoWarn.

Also, "InstrumentKinds" above will revert to "Instrument" to match the spelling of the command-line option.

File: src/Compilers/Core/MSBuildTaskTests/CscTests.cs (hunk 27 above)
Comment (JohnHamby) [inline src/Compilers/Core/MSBuildTaskTests/CscTests.cs]: Adding a test that specifies multiple kinds exposed a bug in how the option was generated. Fixed in the next commit.

**Comment by JohnHamby** [comment]
I will add tests for multiple valid duplicate values for CSharp and for VB.
